### PR TITLE
Implement jcmd

### PR DESCRIPTION
**Pull request by pdbain-ibm**
Add front-end class.  Add help functions for both the utility per se and the target commands.
Add test cases.

This is the first part which adds the framework for running diagnostic
commands. It provides the capabailities of jps, jmap, and jstack via the jcmd
utility.  Later changesets will add further commands using this framework and
more options for the front-end command.

Depends on launchers; pull requests for same:

- [ ] https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/323
- [ ] https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/198
- [ ] https://github.com/ibmruntimes/openj9-openjdk-jdk12/pull/67
- [ ] https://github.com/ibmruntimes/openj9-openjdk-jdk13/pull/7
- [ ] https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/127

Signed-off-by: Peter Bain <peter_bain@ca.ibm.com>

### CODE REVIEW

File: jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+/**
+ * JMap A tool for running diagnostic commands on another Java process
+ *
+ */
+public class Jcmd {
+
+	@SuppressWarnings("nls")
+	private static final String HELPTEXT = "jcmd: run diagnostic command on another Java process%n"
+			+ " Format:%n"
+			// TODO: add the <class> option	+ "    jcmd <vmid | class> <arguments>%n"
```
Comment (pshipton) [inline jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java]: I don't think we should merge a TODO. You can open an Issue is there is some future work to complete.

**Comment by pshipton** [comment]
Please update the description to list the PRs for the launchers.

File: jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java (hunk 1 above)
Comment (pdbain-ibm) [reply jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java]: Done.

**Comment by pdbain-ibm** [comment]
I have listed the pull requests for the launchers.  I believe we can merge the launchers prior to the OpenJ9 code, right @keithc-ca ?

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/AttachApiTest.java [hunk 2]
```diff
@@ -200,6 +203,8 @@ protected boolean checkLoadAgentOutput(String errOutput,
 		ArrayList<String> cmdLine = new ArrayList<>();
 		cmdLine.add(commandName);
 		cmdLine.addAll(args);
+		Optional<String> commandLineString = cmdLine.stream().reduce((s1, s2) -> s1 + " " + s2);
+		log("command line: " + commandLineString.get());
```
Comment (keithc-ca) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/AttachApiTest.java]: `String.join()` is what you want here.
```
log("command line: " + String.join(" ", cmdLine));
```
Alternatively, there's the collector: `joining(String delimiter, String prefix, String suffix)`
```
log(cmdLine.stream().collect(Collectors.joining(" ", "command line: ", "")));
```

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/AttachApiTest.java [hunk 3]
```diff
@@ -212,6 +217,17 @@ protected boolean checkLoadAgentOutput(String errOutput,
 		return outputLines;
 	}
 
+	protected List<String> runCommandAndLogOutput(List<String> commandLineOptions) throws IOException {
+		List<String> jpsOutput = runCommand(commandLineOptions);
+		StringWriter buff = new StringWriter();
+		PrintWriter buffWriter = new PrintWriter(buff);
+		jpsOutput.forEach(s -> {
+			buffWriter.println(s);
+		});
```
Comment (keithc-ca) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/AttachApiTest.java]: In general, there should be a call to `buffWriter.flush()` here to be sure everything is written to `buff`.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		@SuppressWarnings("nls")
+		String[] options = {"", LockedSyncsOption, LockedSyncsOption + "=true", LockedSyncsOption + "=false"};
+		for (String option : options) {
+			List<String> args = new ArrayList<>();
+			args.add(getVmId());
+			args.add(THREAD_PRINT);
+			if (!option.isEmpty()) {
+				args.add(option);
+			}
+			boolean addSynchronizers = !option.endsWith("false") && !option.isEmpty(); //$NON-NLS-1$
```
Comment (keithc-ca) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: Suggest: `boolean addSynchronizers = option.endsWith("=true");`

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+	public void testClassHistogramAll() throws IOException {
+		List<String> args = new ArrayList<>();
+		args.add(getVmId());
+		args.add(GC_CLASS_HISTOGRAM);
+		args.add("-all"); //$NON-NLS-1$
+		List<String> jpsOutput = runCommandAndLogOutput(args);
+		String expectedString = commandExpectedOutputs.getOrDefault(GC_CLASS_HISTOGRAM, "Test error: expected output not defined"); //$NON-NLS-1$
+		log("Expected string: " + expectedString); //$NON-NLS-1$
+		Optional<String> searchResult = StringUtilities.searchSubstring(expectedString, jpsOutput);
+		assertTrue(searchResult.isPresent(), "Expected string not found"); //$NON-NLS-1$
```
Comment (keithc-ca) [inline test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: The assertion message would be more helpful if it mentioned the string that we expected to find.

File: jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.List;
+
+import com.sun.tools.attach.VirtualMachineDescriptor;
+import com.sun.tools.attach.spi.AttachProvider;
+
+import openj9.tools.attach.diagnostics.attacher.AttacherDiagnosticsProvider;
+import openj9.tools.attach.diagnostics.base.DiagnosticUtils;
+
+/**
+ * JMap A tool for running diagnostic commands on another Java process
```
Comment (keithc-ca) [inline jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java]: typo: JMap -> Jcmd

**Comment by keithc-ca** [comment]
> I believe we can merge the launchers prior to the OpenJ9 code, right @keithc-ca ?

Yes. The launcher additions have all been merged.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java (hunk 4 above)
Comment (pdbain-ibm) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/TestJcmd.java]: Unfortunately no: `-l` implies `-l=true`.

File: test/functional/Java8andUp/src/org/openj9/test/attachAPI/AttachApiTest.java (hunk 2 above)
Comment (pdbain-ibm) [reply test/functional/Java8andUp/src/org/openj9/test/attachAPI/AttachApiTest.java]: Thanks for the suggestion.  Done.

File: jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+	@SuppressWarnings("nls")
+	private static final String HELPTEXT = "jcmd: run diagnostic command on another Java process%n"
+			+ " Format:%n"
+			+ "    jcmd <vmid> <arguments>%n"
+			+ " Arguments:%n"
+			+ "    <vmid>: Attach API VM ID as shown in jps or other Attach API-based tools.%n"
+			+ "    command[command arguments]: command from the list returned by \"-help\"%n"
+			+ "    -help: print the list of diagnostic commands%n"
+			+ "    -J: supply arguments to the Java VM running jps%n"
```
[the remark this answers is not in the document]
Comment (pdbain-ibm) [reply jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java]: Oops!

File: jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Util.java [hunk 8]
```diff
@@ -69,4 +78,66 @@ static void runCommandAndPrintResult(AttacherDiagnosticsProvider diagProvider, S
...
+						outputBuffer.append(' ');
+						outputBuffer.append(jvmArguments);
+					}
+				}
+			} finally {
+				theVm.detach();
+			}
+		} catch (AttachNotSupportedException | IOException e) {
+			outputBuffer.append(' ');
+			outputBuffer.append("<no information available>"); //$NON-NLS-1$
```
[the remark this answers is not in the document]
Comment (pdbain-ibm) [reply jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Util.java]: Good plan.

File: jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+			if (null == myProvider) {
+				System.err.println("no attach providers available"); //$NON-NLS-1$
+			} else {
+				for (VirtualMachineDescriptor vmd : myProvider.listVirtualMachines()) {
+					StringBuilder outputBuffer = new StringBuilder(vmd.id());
+					Util.getTargetInformation(myProvider, vmd, false, false, true, outputBuffer);
+					System.out.println(outputBuffer.toString());
+				}
+			}
+		} else if ((args.length == 1) && Arrays.stream(HELP_OPTIONS).anyMatch(s -> s.equals(firstArgument))) {
```
[the remark this answers is not in the document]
Comment (pdbain-ibm) [reply jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java]: I assume `firstArgument` cannot be null because it is created by the launcher.

File: jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+				}
+			}
+		} else if ((args.length == 1) && Arrays.stream(HELP_OPTIONS).anyMatch(s -> s.equals(firstArgument))) {	
+			System.out.printf(HELPTEXT);
+		} else {
+			String vmid = firstArgument;
+			AttacherDiagnosticsProvider diagProvider = new AttacherDiagnosticsProvider();
+			command = DiagnosticUtils.makeJcmdCommand(args, 1);
+			try {
+				diagProvider.attach(vmid);
```
[the remark this answers is not in the document]
Comment (pdbain-ibm) [reply jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java]: `attach` and `runCommandAndPrintResult` can throw an IOE. `detach` is a null operation if the attach fails. I structured it this way to avoid redundant calls to `handleCommandException`, but  I'm open to suggestions for structuring this better.

File: jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Util.java [hunk 11]
```diff
@@ -69,4 +78,66 @@ static void runCommandAndPrintResult(AttacherDiagnosticsProvider diagProvider, S
...
+		final String msg = e.getMessage();
+		if (null == msg) {
+			System.err.println();
+		} else {
+			if (msg.matches(IPC.INCOMPATIBLE_JAVA_VERSION)) {
+				System.err.println(": incompatible target JVM"); //$NON-NLS-1$
+			} else {
+				System.err.printf(": %s%n", msg); //$NON-NLS-1$
+			}
+		}
```
[the remark this answers is not in the document]
Comment (pdbain-ibm) [reply jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Util.java]: Nice idea.  Thanks.

**Comment by pdbain-ibm** [comment]
Thanks for your comments, @keithc-ca .  I have updated per most of your suggestions, though there are a couple of comments where I have responded and would appreciate your follow-up.
When you have no further comments I will re-build and re-test manually.

File: jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (keithc-ca) [reply jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java]: Normally, yes, but if you want to be paranoid, you can ensure it's non-null here.

File: jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (keithc-ca) [reply jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java]: You can wrap the `runCommandAndPrintResult()` call with a `try` whose `finally` calls `detach()`.

[the remark this answers is not in the document]
Comment (pdbain-ibm) [reply jcl/src/jdk.jcmd/share/classes/openj9/tools/attach/diagnostics/tools/Jcmd.java]: Much nicer.  Thanks.

**Comment by keithc-ca** [comment]
Jenkins test extended plinux,win jdk8,jdk11
